zerver/lib/subscription_info.py

```diff
@@ -621,7 +672,9 @@ def get_stream_id(sub_dict: RawSubscriptionDict) -> int:
 
     for raw_stream_dict in never_subscribed_streams:
         is_public = not raw_stream_dict["invite_only"]
-        if is_public or user_profile.is_realm_admin:
+        can_administer_channel_group_id = raw_stream_dict["can_administer_channel_group_id"]
+        is_channel_admin = can_administer_channel_group_id in user_recursive_group_ids
+        if is_public or user_profile.is_realm_admin or is_channel_admin:
```

Comment:
Can we have a function here which checks access for the stream and use it at all these places? It can take user_recursive_group_ids parameter with type set as set[int] | None and we can compute it in the function itself if it is None? Might also make it easy to change access control later.

We have functions for checking stream access already, so maybe we can update those as well but am not sure.